packages/sanity/src/core/bundles/hooks/useDocumentVersions.tsx

```diff
@@ -0,0 +1,119 @@
+import {type ListenEvent, type ListenOptions} from '@sanity/client'
+import {useCallback, useEffect, useMemo, useState} from 'react'
+import {catchError, of} from 'rxjs'
+import {
+  type BundleDocument,
+  DEFAULT_STUDIO_CLIENT_OPTIONS,
+  getBundleSlug,
+  getPublishedId,
+  useBundles,
+  useClient,
+} from 'sanity'
+
+export interface DocumentPerspectiveProps {
+  documentId: string
+}
+
+export interface DocumentPerspectiveState {
+  data: BundleDocument[] | null
+  error?: Error
+}
+
+const LISTEN_OPTIONS: ListenOptions = {
+  events: ['welcome', 'mutation', 'reconnect'],
+  includeResult: true,
+  visibility: 'query',
+}
+
+/**
+ * Fetches the document versions for a given document
+ * @param props - document Id of the document (might include bundle slug)
+ * @returns - data: document versions, loading, errors
+ * @hidden
+ * @beta
+ */
+export function useDocumentVersions(props: DocumentPerspectiveProps): DocumentPerspectiveState {
```

Comment:
Have you considered here instead of setting the listener, use the `listenQuery` function? 
Also, probably we can reduce the data returned from the query, projecting it to include only the id
 `*[sanity::versionOf("${publishedId}")]{_id}`

Reply:
Sorry if I'm missing something, but this looks like it could be handled with the `listenQuery` observer and simplify this hook. 

```
export function useDocumentVersions(props: DocumentPerspectiveProps): DocumentPerspectiveState {
  const {documentId} = props

  const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
  const {data: bundles} = useBundles()
  // TODO: How to identify documents with multiple dots ids, maybe it will be better to provide the whole document?
  const publishedId = getPublishedId(documentId, documentId.includes('.'))

  const QUERY = `*[sanity::versionOf("${publishedId}")]{_id, _version}`

  const observable = useMemo(
    () => listenQuery(client, QUERY) as Observable<Pick<SanityDocument, '_id' | '_version'>[]>,
    [QUERY, client],
  )

   // This will return the sanity documents that are a version of the published document.
  const versions = useObservable(observable, null)

  const state = versions
    ? versions
        .map((r) => bundles?.find((b) => r._version && getBundleSlug(r._id) === b.slug))
        .filter((b) => b !== undefined)
    : null

  return {
    data: state,
  }
}

```

Reply:
I thought about this too, but I don't think it's important enough to block merging, and that it's something that can be addressed down the line? I think it makes sense for us to do a separate refactor/optimization pass to unify some of the data fetching patterns at a later point.

Reply:
Agree, this suggestion is non blocking, it can be addressed later.

Reply:
I've updated with the suggestion from Pedro on slack :) 
Thanks once more!

Reply:
Spoken on slack but had to drop the update due to time limitations (specifically around tests) created a linear story for it.